### PR TITLE
Add poretools formula to homebrew-science

### PR DESCRIPTION
**Pull request by gawbul**
See issue https://github.com/Homebrew/homebrew-science/issues/2299

### CODE REVIEW

File: poretools.rb [hunk 1]
```diff
@@ -0,0 +1,104 @@
+class Poretools < Formula
+  homepage "http://poretools.readthedocs.org/"
+  url "https://github.com/arq5x/poretools/archive/master.tar.gz"
```
Comment (tdsmith) [inline poretools.rb]: This should use a versioned tarball instead of master.

**Comment by tdsmith** [comment]
The h5py build is failing because cython isn't in PATH; I'd add `ENV.prepend_create_path "PATH", libexec/"vendor/bin"` near the top of `install`.

File: poretools.rb [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+  sha256 "4a557a86d0ae43d39d1d472eea5574870fc53466dc0210352f4ce7e9b33f3aa4"
+  version "0.5.1"
+
+  head do
+    url "https://github.com/arq5x/poretools"
+    sha256 "22c7e7e7e6d7273b2c95ef7f638ed838ab476a8dd28d3470edbb3de785cd7625"
+  end
```
Comment (tdsmith) [inline poretools.rb]: Replace the `head` block with `head "https://github.com/arq5x/poretools.git"`

File: poretools.rb [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+      resource(rp).stage do
+        system "python", *Language::Python.setup_install_args(libexec/"vendor")
+      end
+    end
+
+    # install r dependencies
+    resr.each do |rr|
+      resource(rr).stage do
+        # system "R", "CMD", "INSTALL", "."
+        system "R", "-q", "-e", "install.packages('.', repos='http://cran.rstudio.com', dependencies=TRUE)"
```
Comment (tdsmith) [inline poretools.rb]: What's the difference between these two lines?

Comment (tdsmith) [inline poretools.rb]: Also: this should not install packages directly to a user's R package library; Homebrew shouldn't install things outside the Cellar.

File: poretools.rb [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    ENV.prepend_create_path "PYTHONPATH", libexec+"lib/python2.7/site-packages"
+    system "python", *Language::Python.setup_install_args(libexec)
+
+    bin.install Dir[libexec/"bin/*"]
+    bin.env_script_all_files(libexec+"bin", :PYTHONPATH => ENV["PYTHONPATH"])
+  end
+
+  # execute tests
+  test do
+  end
```
Comment (tdsmith) [inline poretools.rb]: A test that loads and does a fast operation on a trivial data set would be great, just to make sure it links correctly and doesn't segfault on run.

File: poretools.rb (hunk 1 above)
Comment (sjackman) [inline poretools.rb]: `https://github.com/arq5x/poretools/archive/v0.5.1.tar.gz`

**Comment by sjackman** [comment]
http://bot.brew.sh/job/Homebrew%20Science%20Pull%20Requests/2009/version=mavericks/testReport/junit/brew-test-bot/mavericks/install_Homebrew_science_poretools/
```
('ERROR: cython exec failed. Is cython not in the path? ', '[Errno 2] No such file or directory')
error: No such file or directory
```

**Comment by dpo** [comment]
Cython doesn't ship with Python. It must be `pip`-installed. I'm not sure we have a policy on Cython packages. Many Cython authors ship the C files generated from the Cython source so the compiler isn't needed.

**Comment by gawbul** [comment]
This is great! Thanks for the comments all :+1: Will make the relevant updates asap :smile:

File: poretools.rb (hunk 3 above)
Comment (gawbul) [inline poretools.rb]: Which two lines? One is Python dependencies and the other R. I want to try and remove the R dependencies from the upstream package, so this should change eventually. What's the best way to install the R packages for now?

File: poretools.rb (hunk 1 above)
Comment (gawbul) [inline poretools.rb]: Updated my code to fix this.

File: poretools.rb (hunk 2 above)
Comment (gawbul) [inline poretools.rb]: Done.

**Comment by gawbul** [comment]
Currently failing on `mavericks` and `mountain_lion` in the build? Not sure why? Not sure how best to work around the R package installation either?

**Comment by iml** [comment]
> Not sure how best to work around the R package installation either?

Are they build-time dependencies? What if someone wants to use R from CRAN instead of Homebrew? @tdsmith is more R savvy but it might not be feasible (or appropriate) to vendor R packages. One possibility could be to install to Cellar and set `R_LIBS_USER` in the env_script wrappers.

**Comment by sjackman** [comment]
> One possibility could be to install to Cellar and set R_LIBS_USER in the env_script wrappers.

That seems reasonable to me, and I believe how vendored libraries are usually handled for other scripting languages in Homebrew. I'm not generally a fan of vendoring libraries, but if you want to be able to bottle poretools, I think that may be the only way of doing it.

**Comment by sjackman** [comment]
http://bot.brew.sh/job/Homebrew%20Science%20Pull%20Requests/2020/version=mavericks/console
`brew test` failure on Mavericks. I don't have access to a Mavericks machine, so can't test this issue further. The error message is not very helpful.
```
==> brew test --verbose Homebrew/science/poretools			 FAILED
Testing poretools
Error: poretools: failed
<"total reads\t297\ntotal base pairs\t260131\nmean\t875.86\nmedian\t795\nmin\t325\nmax\t3602\nN25\t965\nN50\t830\nN75\t741\n"> expected but was
<"">.
```

**Comment by sjackman** [comment]
You could download the bottle from bintray and try running it. Even on a Yosemite machine it might reproduce the error.

Update: There doesn't seem to be any bottles on bintray. https://bintray.com/homebrew/bottles-science/poretools/view#files

**Comment by sjackman** [comment]
If you don't necessarily want bottles, then I'm happy just installing the files that are distributed by upstream, and give the user instructions in `caveats` on which R packages they should install and a shell command to install them.

File: poretools.rb (hunk 3 above)
Comment (tdsmith) [inline poretools.rb]: I meant `R CMD INSTALL` vs `install.packages`.

**Comment by tdsmith** [comment]
> What if someone wants to use R from CRAN instead of Homebrew?

I suggest that in that event we invite them to either send a pull request for a `:r` dependency or pound sand. :)

Bottles are not publicly accessible from bintray until they are published by `brew pull`.

**Comment by sjackman** [comment]
> Bottles are not publicly accessible from bintray until they are published by brew pull.

Ah, I wasn't logged in. Yes, there are three bottles for poretools on bintray. The Yosemite bottle works for me.

```
❯❯❯ brew install ~/Downloads/poretools-0.5.1.yosemite.bottle.tar.gz                 
❯❯❯ /usr/local/Cellar/poretools/0.5.1/bin/poretools stats /usr/local/Cellar/poretools/0.5.1/share/test_data
total reads	297
total base pairs	260131
mean	875.86
median	795
min	325
max	3602
N25	965
N50	830
N75	741
```

File: poretools.rb [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+    min	325
+    max	3602
+    N25	965
+    N50	830
+    N75	741
+    EOS
+
+    result = File.open(testpath/"stats.out").read
+    require "open3"
+    Open3.popen3(bin/"poretools", "stats", share/"test_data/") do |_, stdout, _|
```
Comment (sjackman) [inline poretools.rb]: Can this test instead use:
```ruby
assert_equal result, shell_output("#{bin}/poretools stats #{share}/test_data/
```

**Comment by gawbul** [comment]
Not sure why build failed? Didn't seem related to my changes though (or at least it didn't get as far as installing or testing the formula)?

**Comment by tdsmith** [comment]
Just restarted the build.

**Comment by sjackman** [comment]
http://bot.brew.sh/job/Homebrew%20Science%20Pull%20Requests/2065/version=yosemite/console
```
==> brew readall --syntax						 FAILED
Error: problem in /usr/local/Library/Taps/homebrew/homebrew-science/poretools.rb
==> brew uses Homebrew/science/poretools				Error: /usr/local/Library/Taps/homebrew/homebrew-science/poretools.rb:123: syntax error, unexpected keyword_end, expecting end-of-input
 FAILED
Error: /usr/local/Library/Taps/homebrew/homebrew-science/poretools.rb:123: syntax error, unexpected keyword_end, expecting end-of-input
```

File: poretools.rb [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+    min	325
+    max	3602
+    N25	965
+    N50	830
+    N75	741
+    EOS
+
+    result = File.open(testpath/"stats.out").read
+    assert_equal result, shell_output("#{bin}/poretools stats #{share}/test_data/")
+    end
```
Comment (sjackman) [inline poretools.rb]: This `end` should not be here. Please run `brew audit poretools --strict` before submitting.

**Comment by gawbul** [comment]
Sorry! My bad! Removed `end` and added `desc` in line with `brew edit` suggestions.

**Comment by gawbul** [comment]
Still need to sort the R package installation (which are runtime requirements I believe). Will probably give `vendor`ing the packages a try.

**Comment by gawbul** [comment]
Builds and bottles fine now on `yosemite`, but fails due to missing package `six` on `mavericks` and `mountainlion`.

File: poretools.rb [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+    mean	875.86
+    median	795
+    min	325
+    max	3602
+    N25	965
+    N50	830
+    N75	741
+    EOS
+
+    result = File.open(testpath/"stats.out").read
```
Comment (tdsmith) [inline poretools.rb]: Why round-trip this through a file?

Comment (gawbul) [inline poretools.rb]: Good point! I was just following the example formula file. I've changed this to read into a variable instead. Working on installing the R packages to a vendor library directory too.

Comment (gawbul) [inline poretools.rb]: Trying to run the following command `R -q -e "library(devtools); install_deps(c('/Users/stevemoss/Downloads/ggplot2'), lib='/usr/local/Cellar/poretools/0.5.1/libexec/vendor/library', repos='http://cran.rstudio.com')"`

but get this and don't really understand why:

```
> library(devtools); install_deps(c('/Users/stevemoss/Downloads/ggplot2'), lib='/usr/local/Cellar/poretools/0.5.1/libexec/vendor/library', repos='http://cran.rstudio.com')
Installing 10 packages: colorspace, dichromat, gtable, labeling, munsell, plyr, proto, RColorBrewer, reshape2, scales
Warning: unable to access index for repository source
Warning message:
packages ‘colorspace’, ‘dichromat’, ‘gtable’, ‘labeling’, ‘munsell’, ‘plyr’, ‘proto’, ‘RColorBrewer’, ‘reshape2’, ‘scales’ are not available (for R version 3.2.0)
```

Comment (gawbul) [inline poretools.rb]: Nevermind, fixed it, didn't like having a repos URL too.

**Comment by gawbul** [comment]
Running `$ poretools hist ./MinION_data_1/pass/*.fast5` is giving me:

```Error in loadNamespace(name) : there is no package called ‘ggplot2’
Traceback (most recent call last):
  File "/usr/local/Cellar/poretools/0.5.1/libexec/bin/poretools", line 9, in <module>
    load_entry_point('poretools==0.5.1', 'console_scripts', 'poretools')()
  File "/usr/local/Cellar/poretools/0.5.1/libexec/lib/python2.7/site-packages/poretools/poretools_main.py", line 401, in main
    args.func(parser, args)
  File "/usr/local/Cellar/poretools/0.5.1/libexec/lib/python2.7/site-packages/poretools/poretools_main.py", line 24, in run_subtool
    import hist as submodule
  File "/usr/local/Cellar/poretools/0.5.1/libexec/lib/python2.7/site-packages/poretools/hist.py", line 4, in <module>
    import rpy2.robjects.lib.ggplot2 as ggplot2
  File "/usr/local/Cellar/poretools/0.5.1/libexec/vendor/lib/python2.7/site-packages/rpy2/robjects/lib/ggplot2.py", line 55, in <module>
    ggplot2 = importr('ggplot2')
  File "/usr/local/Cellar/poretools/0.5.1/libexec/vendor/lib/python2.7/site-packages/rpy2/robjects/packages.py", line 399, in importr
    env = _get_namespace(rname)
rpy2.rinterface.RRuntimeError: Error in loadNamespace(name) : there is no package called ‘ggplot2’
```

I assume this is because it isn't picking up the R vendor'd packages. How to modify though?

**Comment by gawbul** [comment]
Is it worth adding a caveats for the R_LIBS_USER?

```
def caveats
  <<-EOS.undent
  Before you can use poretools you must export some variables to your $SHELL.
  export R_LIBS_USER="/usr/local/Cellar/poretools/0.5.1/libexec/vendor/library"
  EOS
end
```

**Comment by tdsmith** [comment]
Whatever environment variable needs to be set (and hopefully that's a viable approach) should be added to the `write_env_script` invocation.

**Comment by gawbul** [comment]
Awesome, thanks! Will update that now! Is there a tutorial anywhere that tells you when and how to use these wrappers?

**Comment by gawbul** [comment]
I think that's it all sorted? Let me know if you spot any issues?

**Comment by gawbul** [comment]
Bah, no, failing because one of the R libraries is linking against the system `openssl` now :disappointed: 

```
==> brew audit Homebrew/science/poretools --strict			 FAILED
==> brew style poretools

1 file inspected, no offenses detected
Error: 1 problems in 1 formulae

==> audit problems
homebrew/science/poretools:
 * object files were linked against system openssl
These object files were linked against the deprecated system OpenSSL.
Adding `depends_on "openssl"` to the formula may help.
  /usr/local/Cellar/poretools/0.5.1/libexec/vendor/R/library/git2r/libs/git2r.so
```

**Comment by gawbul** [comment]
**Now**, I think that is it! Let me know if there are any concerns/issues/thoughts?

**Comment by gawbul** [comment]
Can someone retry the build? Don't think that failure was to do with me? Was it?

**Comment by sjackman** [comment]
@gawbul Click `Details` beside `Failed — Build finished.` See http://bot.brew.sh/job/Homebrew%20Science%20Pull%20Requests/2190/version=yosemite/testReport/junit/brew-test-bot/yosemite/audit_Homebrew_science_poretools___strict/
```
==> brew style poretools

1 file inspected, no offenses detected
Error: 1 problem in 1 formula

==> audit problems
homebrew/science/poretools:
 * `desc`(line 5) should be put before `homepage`(line 2)
```

**Comment by tdsmith** [comment]
This looks good; pulling with a few minor changes (including the desc fix)! Thanks for your patience. It's people like you that help Homebrew thrive.

Just for reference, our preferred commit message format is like "somepackage 1.2.3 (new formula)" for new formulas and we like the commits squashed. I've done that locally.

**Comment by sjackman** [comment]
Thanks for your contribution, Steve, and sticking with a difficult formula.

**Comment by gawbul** [comment]
My pleasure, and thanks for all the input and advice! Is there a page anywhere with details on best practices/how to contribute? I.e. the preferred commit message etc? I guess there's something over in the main `homebrew` repo somewhere, but is there one specific to `homebrew-science` too?

**Comment by iml** [comment]
https://github.com/Homebrew/homebrew-science/blob/master/CONTRIBUTING.md, also has links to the main Homebrew docs.

**Comment by gawbul** [comment]
Lol, I must be blind :wink: Thanks very much :smile:

**Pull request closed without merging**